rust/src/ldap/types.rs

```diff
@@ -0,0 +1,226 @@
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy)]
+pub struct ResultCode(pub u32);
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy)]
+pub struct MessageID(pub u32);
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub struct SearchScope(pub u32);
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub struct DerefAliases(pub u32);
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct LdapString(pub String);
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct LdapDN(pub String);
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct RelativeLdapDN(pub String);
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct LdapOID(pub String);
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct LdapResult {
+    pub result_code: ResultCode,
+    pub matched_dn: LdapDN,
+    pub diagnostic_message: LdapString,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct BindRequest {
+    pub version: u8,
+    pub name: LdapDN,
+    pub authentication: AuthenticationChoice,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct SaslCredentials {
+    pub mechanism: LdapString,
+    pub credentials: Option<Vec<u8>>,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum AuthenticationChoice {
+    Simple(Vec<u8>),
+    Sasl(SaslCredentials),
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct BindResponse {
+    pub result: LdapResult,
+    pub server_sasl_creds: Option<Vec<u8>>,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct SearchRequest {
+    pub base_object: LdapDN,
+    pub scope: SearchScope,
+    pub deref_aliases: DerefAliases,
+    pub size_limit: u32,
+    pub time_limit: u32,
+    pub types_only: bool,
+    pub filter: Filter,
+    pub attributes: Vec<LdapString>,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct SearchResultEntry {
+    pub object_name: LdapDN,
+    pub attributes: Vec<PartialAttribute>,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum ProtocolOp {
+    BindRequest(BindRequest),
+    BindResponse(BindResponse),
+    UnbindRequest,
+    SearchRequest(SearchRequest),
+    SearchResultEntry(SearchResultEntry),
+    SearchResultDone(LdapResult),
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct LdapMessage {
+    pub message_id: MessageID,
+    pub protocol_op: ProtocolOp,
+    pub controls: Option<Vec<Control>>,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Control {
+    pub control_type: LdapOID,
+    pub criticality: bool,
+    pub control_value: Option<Vec<u8>>,
+}
+
+impl<'a> TryFrom<ldap_parser::ldap::LdapMessage<'a>> for LdapMessage {
+    type Error = &'static str;
+
+    fn try_from(ldap_msg: ldap_parser::ldap::LdapMessage) -> Result<Self, Self::Error> {
+        let message_id = MessageID(ldap_msg.message_id.0);
+        let protocol_op = match ldap_msg.protocol_op {
+            ldap_parser::ldap::ProtocolOp::BindRequest(msg) => Self::from_bind_request(msg),
+            ldap_parser::ldap::ProtocolOp::UnbindRequest => ProtocolOp::UnbindRequest,
```

Comment:
Why is there no param for unbind ?

Reply:
Because there is no param for unbind operation.